lib/Dialect/TritonGPU/IR/Dialect.cpp

```diff
@@ -366,15 +366,29 @@ SmallVector<int64_t> getShapePerCTA(ArrayRef<unsigned> CTASplitNum,
   return shapePerCTA;
 }
 
-SmallVector<int64_t> getShapePerCTA(Attribute layout, ArrayRef<int64_t> shape) {
-  if (mlir::isa<SharedEncodingTrait>(layout)) {
+SmallVector<int64_t> getShapePerCTA(Attribute layout,
+                                    ArrayRef<int64_t> shapeOrig) {
+  SmallVector<int64_t> shape(shapeOrig);
+  if (auto sharedMMALayout = mlir::dyn_cast<NVMMASharedEncodingAttr>(layout)) {
+    if (sharedMMALayout.getFp4Padded()) {
+      auto packedAxis = getOrder(sharedMMALayout)[0];
+      if (shape.size() == 3) {
+        // Take into account multi buffering
+        shape[1 + packedAxis] *= 2;
+      } else {
+        shape[packedAxis] *= 2;
```

Comment:
why do we have to multiply the shape by 2 here? Is it for the allocation? I think this logic should live in the allocation code, `getShapePerCTA` should return a logical shape

Reply:
Yes, `getShapePerCTA` is used in the allocation code here https://github.com/triton-lang/triton/blob/main/lib/Analysis/Allocation.cpp#L237

It's also used in `MMAv5.cpp` to set the size of operands. Multiplying the size by two there is important for deciding the right swizzling size and make the SMEM address update work. 

So I thought having `getShapePerCTA` return the physical shape is the best solution. Would there be any code broken if  `getShapePerCTA` doesn't return logical shape?

Reply:
I think so, this helper is commonly used so changing the meaning of it would be a problem.

Reply:
What do you think about introducing a new helper like `gpu::getPhysicalShapePerCTA()` and use in `Allocation.cpp`? I want to avoid repeating the logic of "check if fp4Padded and multiply by two".

Reply:
yeah maybe something like `getPaddedShapePerCTA`?

Reply:
I mean, I want to use that everywhere which expects physical shape, like `Allocation.cpp` or MMA lowering code. So I don't want this new utility to be tied to the padding stuff.

Reply:
I think the concept of physical shape is confusing as the data are swizzled in memory so they are technically higher dimension. I would prefer a name where we consider a bigger shape because we are padding out the data.
Otherwise maybe `getAllocationShapePerCTA()`?

Reply:
yeah I like `getAllocationShapePerCTA()`

Reply:
Done, it took me a while to find one last use of `getShapePerCTA` that needs to be replaced with `getAllocationShapePerCTA()`. It's in `SharedMemoryObject::getStrides()`, it literally asks for `allocShapePerCTA` :)